refactor(agw): Use netifaces to get mac addresses consistently

MoritzThomasHuebner opened the pull request:
Closes #14298

With @wolfseb 

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary

- Remove redundant `get_mac_address` from `bridge_util.py`
- Unified the way we get mac addresses from interfaces
- Added some more tests for `ifaces.py`
- Update Bazel stuff
- Change mocks in tests to make work

<!-- Enumerate changes you made and why you made them -->

## Test Plan

- [x] Run CI tests

<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

MoritzThomasHuebner committed:
refactor(agw): Introduce getmac.get_mac_address

- Replace netifaces in `magma.pipelined.ifaces.get_mac_address` with getmac
- Remove redundant `get_mac_address` from bridge_util.py
- Replace scapy `getmacbyip` and `getmacbyip6` calls in ebpf_manager.py and gw_mac_address.py
- Adjusted ifaces.py tests
- Update Bazel requirements and BUILD.bazel

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Introduce getmac.get_mac_address

- Replace netifaces in `magma.pipelined.ifaces.get_mac_address` with getmac
- Remove redundant `get_mac_address` from bridge_util.py
- Replace scapy `getmacbyip` and `getmacbyip6` calls in ebpf_manager.py and gw_mac_address.py
- Adjusted ifaces.py tests
- Update Bazel requirements and BUILD.bazel

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
Added ifaces to epbf_manager dependencies

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

sebathomas commented on lte/gateway/python/magma/pipelined/bridge_util.py:
```diff
@@ -96,22 +96,6 @@ def get_ofport(interface_name):
         return int(port_num)
 
-    @staticmethod
-    def get_mac_address(interface_name):
```
There's another function `get_virtual_iface_mac` in `utils.py` that you can probably delete.

sebathomas commented on lte/gateway/python/magma/pipelined/tests/test_ifaces.py:
```diff
@@ -22,17 +22,14 @@
-def test_get_mac_address_invalid(netifaces_mock):
-    netifaces_mock.AF_LINK = 13
-    netifaces_mock.ifaddresses.return_value = {netifaces_mock.AF_LINK: []}
+@patch("magma.pipelined.ifaces.get_mac_address")
```
This looks like it's mocking the function that you're trying to test. I think you need to mock `getmac` instead.

MoritzThomasHuebner replied:
We removed this files

sebathomas commented on lte/gateway/python/magma/pipelined/ifaces.py:
```diff
@@ -31,8 +32,13 @@ def monitor_ifaces(iface_names):
-    if not if_addresses or not if_addresses[0].get('addr'):
-        raise ValueError(f"No mac address found for interface {interface_name}")
-    return if_addresses[0]['addr']
+def get_mac_address(network_request=True, **kwargs) -> str:
```
I wonder if it even makes sense to keep this wrapper instead of just using `getmac` directly. Adding a wrapper just for some extra error handling and argument validation is usually not worth it.

(If you call it directly, you can also remove the corresponding unit test.)

MoritzThomasHuebner replied:
We removed the wrapper

Commit:
refactor(agw): replace get_virtual_iface_mac with get_mac_address

* replace function calls to utils/get_virtual_iface_mac with
  ifaces/get_mac_address
* adapt mocks in sudo tests to work with the changed function calls

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
refactor(agw): replace get_virtual_iface_mac with get_mac_address

* replace function calls to utils/get_virtual_iface_mac with
  ifaces/get_mac_address
* adapt mocks in sudo tests to work with the changed function calls

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Make tests green again

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Make tests green again

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

nstng commented on lte/gateway/python/magma/pipelined/ebpf/BUILD.bazel:
```diff
@@ -22,6 +22,7 @@ py_library(
+        requirement("getmac"),
         requirement("netifaces"),
         requirement("pyroute2"),
         requirement("scapy"),
```
scapy is not used here anymore, right?

MoritzThomasHuebner replied:
Yes, I will remove it

jheidbrink commented on lte/gateway/python/magma/pipelined/utils.py:
```diff
@@ -86,8 +85,3 @@ def get_ovs_bridge(cls, datapath):
         return cls.ovs
-
-
-def get_virtual_iface_mac(iface):
```
What was wrong with this one?

MoritzThomasHuebner replied:
There is nothing wrong with it, we are just trying to unify the way mac addresses are inferred in magma. Previously, we used `get_virtual_iface_mac` which depends on `netifaces` to get the mac from a network interface, and various `scapy` utilities gets the mac address from an IP. There was `get_mac_address` in `bridge_utils.py` which used some bash commands to get the mac from the network interface name.  Also, there was yet another utility in `ifaces.py`, which also relies on the same `netifaces` logic as `get_virtual_iface_mac`. As I am writing, I also discovered two other local `get_virtual_iface_mac` functions in `check_quota.py` and `redirect.py` which are identical to this one. It's not really clear why they should exist. We want to simplify this by using `getmac` everywhere.

jheidbrink commented on lte/gateway/python/magma/pipelined/BUILD.bazel:
```diff
@@ -247,7 +248,11 @@ py_library(
+    deps = [
+        ":ifaces",
+        requirement("getmac"),
+        requirement("scapy"),
```
Why did this work previously? Didn't it require scapy before?

sebathomas replied:
It's a bit surprising. Maybe it's only imported in a context where scapy is also imported via mandatory_apps, where it was removed now?

jheidbrink replied:
Right, `mandatory_apps` is the only reverse-dependency of `gw_mac_address`.

nstng commented on lte/gateway/python/magma/pipelined/app/arp.py:
```diff
@@ -13,13 +13,13 @@
 
 from collections import namedtuple
 
+import getmac
```
holds for several files: I suggest to use a from-import

MoritzThomasHuebner replied:
Unfortunately, this messes with mocks used elsewhere since python mocks based on names. In related test code I would have to write something like
```
@patch("magma.pipelined.app.arp.get_mac_address", return_value=...)
```
Instead of the perhaps more obvious
```
@patch("getmac.get_mac_address", return_value=...)
```

MoritzThomasHuebner committed:
refactor(agw): Removing some more local functions

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Removing some more local functions

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Update Bazel

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

crasu reviewed (changes requested):
We probably want to use another pkg then getmac [address](https://github.com/magma/magma/pull/14355#issuecomment-1311357636).

Commit:
refactor(agw): switch to custom get_mac_address function

Remove getmac.get_mac_address and use a custom cuntion that uses
netifaces to find the mac address from an interface name, IPv4, or
IPv6 address

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
refactor(agw): switch to custom get_mac_address function

Remove getmac.get_mac_address and use a custom cuntion that uses
netifaces to find the mac address from an interface name, IPv4, or
IPv6 address

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
refactor(agw): introduce test for get_mac_address function

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
refactor(agw): introduce test for get_mac_address function

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Replaced getmac with netifaces

- Updated Bazel files
- Updated requirements.in
- Updated various tests and mocks
- Fixed an incorrect keyword in a `get_mac_address` call

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Replaced getmac with netifaces

- Updated Bazel files
- Updated requirements.in
- Updated various tests and mocks
- Fixed an incorrect keyword in a `get_mac_address` call

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

sebathomas commented on lte/gateway/python/magma/pipelined/ifaces.py:
```diff
@@ -31,8 +32,33 @@ def monitor_ifaces(iface_names):
+    raise ValueError("Must specify either interface or ip4 or ip6")
+
+
+def _get_mac_from_ip(ip: str, ipv: str) -> str:
```
It seems the function we're trying to replace was also able to find addresses for remote IPs. Do we need that feature?

wolfseb replied:
Isn't that what `get_gw_mac_address` is implementing?

sebathomas commented on lte/gateway/python/magma/pipelined/ifaces.py:
```diff
@@ -31,8 +32,33 @@ def monitor_ifaces(iface_names):
+def get_mac_address(
+        interface: Optional[str] = None,
+        ip4: Optional[str] = None,
+        ip6: Optional[str] = None,
```
Why not just make it two or three methods?

wolfseb replied:
Done.

crasu commented on lte/gateway/python/magma/pipelined/ebpf/ebpf_manager.py:
```diff
@@ -306,22 +305,23 @@ def _get_if_name(self, if_index: int):
         LOG.debug("if-name: %s, mac: %s" % (if_name, addr_str))
         return self._pack_mac_addr(addr_str)
 
     def _get_mac_address_of_ip(self, ip_addr: IPAddress):
```
These gw ips could be remote ips - Needs more investigation. These IPs come from mobilityD via GetGatewayInfo.

crasu replied:
After looking at this again - this was using netifaces before so this should not be dangerous.

sebathomas replied:
Are you sure? To me it seems like it was always scapy in this file.

crasu replied:
Conversation continued here:
https://github.com/magma/magma/pull/14355#discussion_r1022619162

crasu reviewed (commented):
I found one place that looks potentially dangerous

Commit:
refactor(agw): split get_mac_address by input

Introduce the functions `get_mac_address_from_{iface/ip4/ip6} to split
up `get_mac_address` by the type of input (interface name, IPv4 address,
IPv6 address)

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
refactor(agw): rebase

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

sebathomas commented on lte/gateway/python/magma/pipelined/ifaces.py:
```diff
@@ -31,8 +32,31 @@ def monitor_ifaces(iface_names):
-    return if_addresses[0]['addr']
+def get_mac_address_from_iface(iface: str) -> str:
+    ifaddress = netifaces.ifaddresses(iface)
+    if netifaces.AF_LINK in ifaddress:
```
The old code and the new code have different kinds of null checks. Do we know which of these cases can actually happen with the library? If not, maybe we should do all of them.
(Just for consistency, it's not that important since we're going to raise an exception in any case.)

wolfseb replied:
Right, actually, we could leave the old code, as we do not change the functionality at all and just rename the function

crasu reviewed (commented):
LGTM

crasu commented on lte/gateway/python/magma/pipelined/ebpf/ebpf_manager.py:
```diff
@@ -306,22 +309,23 @@ def _get_if_name(self, if_index: int):
+        try:
+            if ip_addr.version == IPAddress.IPV4:
+                ip_str = socket.inet_ntop(socket.AF_INET, ip_addr.address)
+                addr_str = get_mac_address_from_ip4(ip_str)
```
@sebathomas These gateway IPs are remote ips. 

We could use the mac addresses coming from mobilityd:
https://github.com/magma/magma/blob/a5d0ba40af4db5c9130cfcf7ea86bb9c4840c2f4/lte/protos/mobilityd.proto#L133

But these are potentially wrong. I would suggest to leave this as is for now and open up a follow up ticket for this case.

crasu replied:
We will handle  this in #14442. Just remove this functionallity for now and leave the old scapy code.

wolfseb replied:
For IP4, this is what `_get_gw_mac_address_v4` is doing. For IPv6, we should probably change back to scapy as well in `_get_gw_mac_address_v6` and add that to the new issue.

crasu replied:
Yepp v4 looks fine - maybe you want to mention this in #14442. ipv6 does not use arp but NDP (neigh).

Commit:
refactor(agw): return to original get_mac_from_iface

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
fix(agw): return to scapy for remote ip6 handling

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Revert some changes

- Make the contents of the PR consistent
- Simplify use of `gw_ip`
- Adjust Bazel stuff

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

pshelar commented on lte/gateway/python/magma/pipelined/ifaces.py:
```diff
@@ -31,8 +31,31 @@ def monitor_ifaces(iface_names):
+
+def get_mac_address_from_ip6(ip6: str) -> str:
+    return _get_mac_from_ip(ip6, "ip6")
+
```
why these functions are added ? I only see its used in test code.

MoritzThomasHuebner replied:
Hi, I think you are are right about this. We removed these features from this PR. We have some follow-up PRs that may need them going forward, but we can just get the changes back if we really need them.

pshelar commented on lte/gateway/python/magma/pipelined/ifaces.py:
```diff
@@ -31,8 +31,31 @@ def monitor_ifaces(iface_names):
+                return ifaddress[netifaces.AF_LINK][0]['addr']
+        except (KeyError, IndexError, ValueError):
+            continue
+    raise ValueError(f"No mac address found for {ipv} {ip}")
```
This function is confusing since in networking neighbor discovery protocol is used for mapping IP address to its mac address.
You could add 'local' in function name to be clear about its functionality.

MoritzThomasHuebner replied:
See my comment above

MoritzThomasHuebner committed:
refactor(agw): Fix iface tests

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Remove get local mac by ip logic

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

pshelar reviewed (approved):
LGTM

nstng reviewed (approved):
lgtm
